feat: [GM-23] [BE] Implement the get package endpoint

krishankumar01 opened the pull request:
Case 1:
![image](https://github.com/goravel-ecosystem/market-backend/assets/84431594/452d30fb-b581-4ae7-a7c0-3ac741f5b5b4)
![image](https://github.com/goravel-ecosystem/market-backend/assets/84431594/168c2b12-3543-4fda-87ce-3b51420db5cc)

Case 2:
![image](https://github.com/goravel-ecosystem/market-backend/assets/84431594/83968895-9221-48b7-a522-1ac7caba6f40)

Case 3:
![image](https://github.com/goravel-ecosystem/market-backend/assets/84431594/a7164d78-5c29-4b5e-9fda-72a91dcc7984)

krishankumar01 committed:
implement GetPackage

Commit:
chore: update doc

hwbrzzl commented on src/doc/index.html:
```diff
@@ -423,9 +423,9 @@ <h4>Methods with HTTP bindings</h4>
                 <td>GetPackage</td>
-                <td>POST</td>
+                <td>GET</td>
                 <td>/business/packages/{id}</td>
-                <td>*</td>
+                <td></td>
```
Do you modify this by the `make doc` command?

hwbrzzl replied:
Oh, sorry, ignore this, I think this modification is committed by github action.

hwbrzzl reviewed (commented):
I don't find the modification of `src/go/proto/packages/*`?

krishankumar01 commented:
> I don't find the modification of `src/go/proto/packages/*`?

I changed 'post' to 'get' and removed the body. Additionally, I didn't push it to GitHub, but I also added some fields in the 'Package' message.
![image](https://github.com/goravel-ecosystem/market-backend/assets/84431594/95b75f78-634b-42ad-a28d-8d297eef6d18)

krishankumar01 committed:
Merge branch 'master' into kkumar-gcc/gm-22

hwbrzzl committed:
Compile package proto

krishankumar01 committed:
add translations

krishankumar01 committed:
Merge remote-tracking branch 'origin/kkumar-gcc/gm-22' into kkumar-gcc/gm-22

krishankumar01 committed:
push package files generated by protoc

krishankumar01 committed:
update vpn readme

hwbrzzl committed:
Update readme

hwbrzzl commented on src/go/gateway/config/grpc.go:
```diff
@@ -20,9 +21,12 @@ func init() {
 		// Interceptors can be the group name of UnaryClientInterceptorGroups in app/grpc/kernel.go.
 		"clients": map[string]any{
 			"user": map[string]any{
```
Need to create a new key: `package`

hwbrzzl commented on src/go/gateway/config/grpc.go:
```diff
@@ -20,9 +21,12 @@ func init() {
-				"host":         config.Env("GRPC_USER_HOST", ""),
-				"port":         config.Env("GRPC_USER_PORT", ""),
-				"handlers":     []gateway.Handler{protouser.RegisterUserServiceHandler},
+				"host": config.Env("GRPC_USER_HOST", ""),
+				"port": config.Env("GRPC_USER_PORT", ""),
```
Use `GRPC_PACKAGE_HOST` and `GRPC_PACKAGE_PORT`.

krishankumar01 replied:
Yeah, most of these comment is already solved. I haven't pushed these changes to github yet.

hwbrzzl commented on src/proto/package/package.proto:
```diff
@@ -35,8 +35,7 @@ message GetPackageResponse {
     option (google.api.http) = {
-      post: "/business/packages/{id}"
-      body: "*"
+      get: "/business/packages/{id}"
```
```suggestion
      get: "/packages/{id}"
```

hwbrzzl commented on src/go/package/app/grpc/controllers/package_controller.go:
```diff
@@ -0,0 +1,54 @@
+	"github.com/goravel/framework/database/orm"
+	"github.com/goravel/framework/facades"
+
+	protopackage "market.goravel.dev/proto/package"
+
+	"market.goravel.dev/package/app/services"
```
```suggestion
	protopackage "market.goravel.dev/proto/package"
	"market.goravel.dev/package/app/services"
```

hwbrzzl commented on src/go/package/lang/cn.json:
```diff
@@ -0,0 +1,9 @@
+    "user_id": "UserID 不能为空"
+  },
+  "not_exist": {
+    "package": "找不到包"
```
```suggestion
    "package": "包不存在"
```

hwbrzzl commented on src/go/package/database/migrations/20240101034640_create_packages_table.up.sql:
```diff
@@ -9,5 +9,5 @@ CREATE TABLE packages (
   created_at timestamp NOT NULL,
   updated_at timestamp NOT NULL,
-  deleted_at timestamp NOT NULL
+  deleted_at timestamp DEFAULT NULL
```
Good catch

hwbrzzl commented on src/go/package/routes/grpc.go:
```diff
@@ -1,5 +1,13 @@
+import (
+	"github.com/goravel/framework/facades"
+
+	packageproto "market.goravel.dev/proto/package"
 
+	"market.goravel.dev/package/app/grpc/controllers"
```
```suggestion
	packageproto "market.goravel.dev/proto/package"
	"market.goravel.dev/package/app/grpc/controllers"
```

hwbrzzl commented on src/proto/package/package.proto:
```diff
@@ -35,8 +35,7 @@ message GetPackageResponse {
 service PackageService {
   rpc GetPackage (GetPackageRequest) returns (GetPackageResponse) {
```
This package needs to return the package tags and the User information.

<img width="900" alt="image" src="https://github.com/goravel-ecosystem/market-backend/assets/24771476/a6878400-d43f-40aa-bed4-3321a6e3f29e">

hwbrzzl replied:
This PR is good, we can leave it here for a while, and implement https://goravel.atlassian.net/browse/GM-25 first, then we can get tags in this PR.

hwbrzzl replied:
Sorry ~ I forgot the relationship between these two tickets when you picked up this ticket.

krishankumar01 replied:
How is GM-25 is related to this issue? There we are searching a tag by its name but here we want to get all related tags to a package.

hwbrzzl replied:
The GetTags endpoint can add a filter field: package_id, it can return all tags of a package. Optimized the description. 

<img width="503" alt="image" src="https://github.com/goravel-ecosystem/market-backend/assets/24771476/b57b6e2f-7385-435b-b617-be11ccd2863b">

hwbrzzl commented on src/go/package/app/models/package_test.go:
```diff
@@ -0,0 +1,145 @@
+
+type PackageSuite struct {
+	suite.Suite
+	pack *Package
```
```suggestion
	package *Package
```

krishankumar01 replied:
package is a keyword.

hwbrzzl commented on src/go/package/app/models/package_test.go:
```diff
@@ -0,0 +1,145 @@
+			},
+		},
+		{
+			name: "Sad path - get user error",
```
```suggestion
			name: "Sad path - get package error",
```

hwbrzzl commented on src/go/package/app/grpc/controllers/status.go:
I think we can move this to `src/utils/response`.

hwbrzzl commented on src/go/package/app/grpc/controllers/package_controller.go:
```diff
@@ -0,0 +1,54 @@
+			Status: NewBadRequestStatus(errors.New(facades.Lang(ctx).Get("required.user_id"))),
+		}, nil
+	}
+	packageId := req.GetId()
```
```suggestion
	packageID := req.GetId()
```

hwbrzzl commented on src/go/package/app/grpc/controllers/package_controller.go:
```diff
@@ -0,0 +1,54 @@
+		}, nil
+	}
+
+	packageData, err := r.packageService.GetPackageByID(packageId)
```
```suggestion
	package, err := r.packageService.GetPackageByID(packageId)
```

hwbrzzl commented on src/go/package/app/grpc/controllers/package_controller_test.go:
```diff
@@ -0,0 +1,38 @@
+}
+
+func (s *PackageControllerSuite) TestGetPackage() {
+	// implement me
```
Just a reminder.

krishankumar01 replied:
Yeah, added a TODO, will finish it later.

hwbrzzl reviewed (commented):
A good PR 👍

krishankumar01 committed:
add a package client in grpc config

Commit:
chore: update doc

krishankumar01 committed:
use utils response into user test

krishankumar01 committed:
Merge branch 'master' into kkumar-gcc/gm-22

krishankumar01 committed:
regenerate proto for package

krishankumar01 committed:
remove a duplicate package key from grpc config

hwbrzzl committed:
Add an example for calling UserService

krishankumar01 committed:
add user and tags to package proto

Commit:
chore: update doc

krishankumar01 committed:
add the test cases for GetPackage

krishankumar01 committed:
zero value check on pkg in GetPackage

krishankumar01 committed:
fix test

krishankumar01 committed:
fix test

krishankumar01 committed:
fix test for user microservice

hwbrzzl commented on src/go/package/app/models/package.go:
```diff
@@ -16,3 +25,32 @@ type Package struct {
+func (r *Package) ToProto() *protopackage.Package {
+	return &protopackage.Package{
+		Id:            cast.ToString(r.ID),
+		UserId:        cast.ToString(r.UserID),
```
FYI, the reason of `Id`, `UserId`, etc return string to FE but are stored by uint64 is FE can't parse the uint64 type, will throw some bits, eg: ID: `1234567890123456789`(BE) -> ID: `1234567890123400000`(FE).

krishankumar01 replied:
Thanks for information.

hwbrzzl commented on src/go/package/app/models/package.go:
```diff
@@ -16,3 +25,32 @@ type Package struct {
+		LastUpdatedAt: r.LastUpdatedAt.ToString(),
+		CreatedAt:     r.CreatedAt.ToString(),
+		UpdatedAt:     r.UpdatedAt.ToString(),
+		DeletedAt:     carbon.FromStdTime(r.DeletedAt.Time).ToString(),
```
Don't need to return DeletedAt to FE.

hwbrzzl commented on src/proto/package/package.proto:
```diff
@@ -20,6 +21,8 @@ message Package {
   string created_at = 9;
   string updated_at = 10;
   string deleted_at = 11;
```
Remove the `delete_at` field.

krishankumar01 committed:
remove deleted_at from proto

Commit:
chore: update doc

hwbrzzl reviewed (approved):
Great PR 👍

hwbrzzl commented:
CI stuck, close and reopen this PR.
